sqldelight-gradle/plugin/src/main/kotlin/com/squareup/sqldelight/gradle/VerifyMigrationTask.kt

```diff
@@ -0,0 +1,124 @@
+import schemacrawler.schemacrawler.SchemaInfoLevelBuilder
+import schemacrawler.utility.SchemaCrawlerUtility
+import java.io.File
+import java.sql.Connection
+import java.sql.DriverManager
+
+open class VerifyMigrationTask : SourceTask() {
+  @Suppress("unused") // Required to invalidate the task on version updates.
+  @Input fun pluginVersion() = VERSION
+
+  lateinit var sourceFolders: Iterable<File>
+
+  private val environment by lazy {
+    SqlDelightEnvironment(sourceFolders = sourceFolders.filter { it.exists() })
+  }
+
+  @TaskAction
+  fun generateSchemaFile() {
+    val currentDb = DriverManager.getConnection("jdbc:sqlite:")
+    val sourceFiles = ArrayList<SqlDelightFile>()
+    environment.forSourceFiles { file -> sourceFiles.add(file as SqlDelightFile) }
+    sourceFiles.forInitializationStatements { sqlText ->
+      currentDb.prepareStatement(sqlText).execute()
+    }
+
+    val folders = sourceFolders.toMutableList()
+    while (folders.isNotEmpty()) {
+      val folder = folders.removeAt(0)
+      for (file in folder.listFiles()) {
+        if (file.name.endsWith(".db")) {
+          checkMigration(file, currentDb)
+        } else if (file.isDirectory) {
+          folders.add(file)
+        }
+      }
+    }
+  }
+
+  private fun checkMigration(dbFile: File, currentDb: Connection) {
+    val schemaCrawlerOptions = SchemaCrawlerOptions()
+    schemaCrawlerOptions.schemaInfoLevel = SchemaInfoLevelBuilder.maximum()
+    val expectedCatalog = SchemaCrawlerUtility.getCatalog(currentDb, schemaCrawlerOptions)
+
+    val version = dbFile.nameWithoutExtension.toLong()
+    val copy = dbFile.copyTo(File("${project.buildDir}/sqldelight/${dbFile.name}"))
+    val connection = DriverManager.getConnection("jdbc:sqlite:${copy.absolutePath}")
+    environment.forMigrationFiles {
+      val migrationVersion = it.virtualFile.nameWithoutExtension.toLong()
+      if (version > migrationVersion) return@forMigrationFiles
+      it.sqlStmtList!!.statementList.forEach {
+        connection.prepareStatement(it.rawSqlText()).execute()
+      }
+    }
+    val actualCatalog = SchemaCrawlerUtility.getCatalog(connection, schemaCrawlerOptions)
+
+    val objectDifferBuilder = SchemaCrawlerDifferBuilder()
+    val diff = objectDifferBuilder.build().compare(expectedCatalog, actualCatalog)
+
+    val sb = StringBuilder()
+    diff.visit { node, visit ->
+      if (CrawlInfo::class.java.isAssignableFrom(node.valueType) ||
+          JdbcDriverInfo::class.java.isAssignableFrom(node.valueType)) {
+        visit.dontGoDeeper()
+        return@visit
+      }
+      sb.append("${node.path} - ${node.state}\n")
+    }
+    // TODO: Figure out how to verify triggers/views/indexes are migrated.
+    try {
+      if (sb.isNotEmpty()) {
+        throw IllegalStateException("Error migrating from ${dbFile.name}:\n$sb")
+      }
+    } finally {
+      copy.delete()
+      connection.close()
+    }
+  }
+}
+
+class SchemaCrawlerDifferBuilder {
```

Comment:
private?

Reply:
It's a copy-paste of a class that should be in one of SchemaCrawler artifacts, but we couldn't figure out which one!